aws-rds-dbinstance/aws-rds-dbinstance.json

```diff
@@ -0,0 +1,427 @@
+    },
+    "AvailabilityZone": {
+      "type": "string",
+      "description": "The Availability Zone (AZ) where the database will be created. For information on AWS Regions and Availability Zones."
+    },
+    "BackupRetentionPeriod": {
+      "type": "integer",
+      "minimum": 0,
+      "maximum": 35,
+      "default": 1,
+      "description": "The number of days for which automated backups are retained. Setting this parameter to a positive number enables backups. Setting this parameter to 0 disables automated backups."
+    },
+    "CACertificateIdentifier": {
+      "type": "string",
+      "description": "The identifier of the CA certificate for this DB instance."
+    },
+    "CharacterSetName": {
+      "type": "string",
+      "description": "For supported engines, indicates that the DB instance should be associated with the specified character set."
+    },
+    "CopyTagsToSnapshot": {
+      "type": "boolean",
+      "description": "A value that indicates whether to copy tags from the DB instance to snapshots of the DB instance. By default, tags are not copied."
+    },
+    "DBClusterIdentifier": {
+      "type": "string",
+      "description": "The identifier of the DB cluster that the instance will belong to.",
+      "pattern": "^[a-zA-Z]{1}(?:-?[a-zA-Z0-9]){0,62}$"
+    },
+    "DBInstanceClass": {
+      "type": "string",
+      "description": "The compute and memory capacity of the DB instance, for example, db.m4.large. Not all DB instance classes are available in all AWS Regions, or for all database engines."
+    },
+    "DBInstanceIdentifier": {
+      "type": "string",
+      "pattern": "^[a-zA-Z]{1}(?:-?[a-zA-Z0-9]){0,62}$",
+      "description": "A name for the DB instance. If you specify a name, AWS CloudFormation converts it to lowercase. If you don't specify a name, AWS CloudFormation generates a unique physical ID and uses that ID for the DB instance."
+    },
+    "DBName": {
+      "type": "string",
+      "pattern": "^[_a-zA-Z][a-zA-Z0-9]{0,63}$",
+      "description": "The meaning of this parameter differs according to the database engine you use."
+    },
+    "DBParameterGroupName": {
+      "type": "string",
+      "pattern": "^[a-zA-Z]{1}(?:-?[a-zA-Z0-9\\.]){0,254}$",
+      "description": "The name of an existing DB parameter group or a reference to an AWS::RDS::DBParameterGroup resource created in the template."
+    },
+    "DBSecurityGroups": {
+      "type": "array",
+      "uniqueItems": true,
+      "items": {
+        "type": "string"
+      },
+      "description": "A list of the DB security groups to assign to the DB instance. The list can include both the name of existing DB security groups or references to AWS::RDS::DBSecurityGroup resources created in the template."
+    },
+    "DBSnapshotIdentifier": {
```

Comment:
this should be modeled via conditional operators - `oneOf`, `allOf`, `anyOf`
```
If you specify the DBSnapshotIdentifier property to restore a DB instance (as opposed to specifying it for DB instance updates), then don't specify the following properties:
    AllocatedStorage
    BackupRetentionPeriod
    CharacterSetName
    DBClusterIdentifier
    DBName
    DBSecurityGroups
    DeleteAutomatedBackups
    EnablePerformanceInsights
    EngineVersion
    KmsKeyId
    MasterUsername
    MasterUserPassword
    MaxAllocatedStorage
    MonitoringInterval
    MonitoringRoleArn
    PerformanceInsightsKMSKeyId
    PerformanceInsightsRetentionPeriod
    PreferredBackupWindow
    PreferredMaintenanceWindow
    PromotionTier
    SourceDBInstanceIdentifier
    SourceRegion
    StorageEncrypted
    Timezone
```

Reply:
@ammokhov could you please elaborate how a negation could be implemented with the current schema?

I tried something like:
```json
{
  "not": {
    "required": [
      "AllocatedStorage",
      "BackupRetentionPeriod",
      ...
      "SourceRegion",
      "StorageEncrypted",
      "Timezone"
    ]
  }
}
```
But this conflicts with the [schema definition](https://schema.cloudformation.us-east-1.amazonaws.com/provider.definition.schema.v1.json).